### DISCUSSION TITLE
Connect-OVMgmt :specified API version is not valid. Check the value of the X-API-Version header, then resend the request

### DISCUSSION THREAD

**Discussion by zeshan1977** [HewlettPackard/POSH-HPOneView#579] [Q&A]
<PRE>

PS C:\dev> Connect-OVMgmt -Hostname 19.XX.XX.XXX  -UserName administrator -Password 'XXXXX' -AuthLoginDomain Local -Verbose
WARNING: Parameter 'UserName' is obsolete. 
WARNING: Parameter 'Password' is obsolete. 

VERBOSE: [CONNECT-OVMGMT] Bound PS Parameters: 
Name                           Value                                                                           
----                           -----                                                                           
UserName                       administrator                                                                   
Verbose                        True                                                                            
Hostname                       19.XX.XX.XXX                                                                    
AuthLoginDomain                Local                                                                           
Password                       [*****REDACTED******]                                                           

VERBOSE: [CONNECT-OVMGMT] Creating Session Container
VERBOSE: [CONNECT-OVMGMT] Checking X-API Version.
VERBOSE: [SEND-OVREQUEST] BEGIN
VERBOSE: [SEND-OVREQUEST] Called from: Connect-OVMgmt
VERBOSE: [SEND-OVREQUEST] Bound PS Parameters: 
Key      Value        
---      -----        
uri      /rest/version
Hostname 19.XX.XXX.XXX

VERBOSE: [SEND-OVREQUEST] Process
VERBOSE: [SEND-OVREQUEST] Hostname value: XXXXXXXXX

VERBOSE: [SEND-OVREQUEST] Filtering for Connection Object via String:XXXXXXXXX
VERBOSE: [SEND-OVREQUEST] Processing 'XXXXXXXXX appliance connection request. 1 of 1
VERBOSE: [SEND-OVREQUEST] Requested URI '/rest/version' to 'XXXXXXXXX'
VERBOSE: [SEND-OVREQUEST] We have reached the URI Whitelist condition block. Unauth request allowed for '/rest/v
ersion'.
VERBOSE: [SEND-OVREQUEST] Request: GET https://XXXXXXXXX/rest/version
VERBOSE: [SEND-OVREQUEST] Request Header 1: User-Agent = HPEOneView.PowerShell/6.10 (Microsoft Windows NT 10.0.1
9042.0)
VERBOSE: [SEND-OVREQUEST] Request Header 2: Content-Type = application/json
VERBOSE: [SEND-OVREQUEST] Request Header 3: Accept = application/json, text/html, application/xhtml+xml
VERBOSE: [SEND-OVREQUEST] Request Header 4: X-API-Version = 2800
VERBOSE: [SEND-OVREQUEST] Request Header 5: accept-language = en_US
VERBOSE: [SEND-OVREQUEST] Request Header 6: accept-encoding = gzip, deflate
VERBOSE: [SEND-OVREQUEST] Request Header 7: auth = [*****REDACTED******]
VERBOSE: [SEND-OVREQUEST] Net.WebException Error caught
VERBOSE: [SEND-OVREQUEST] Exception Object: 

PSMessageDetails      : 
Exception             : System.Net.WebException: The remote server returned an error: (412) Precondition 
                        Failed.
                           at System.Net.HttpWebRequest.GetResponse()
                           at CallSite.Target(Closure , CallSite , WebRequest )
TargetObject          : 
CategoryInfo          : NotSpecified: (:) [], WebException
FullyQualifiedErrorId : WebException
ErrorDetails          : 
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at Send-OVRequest<Process>, C:\Program 
                        Files\WindowsPowerShell\Modules\HPEOneView.610\6.10.2778.2020\HPEOneView.610.psm1: 
                        line 6083
                        at Connect-OVMgmt<Begin>, C:\Program 
                        Files\WindowsPowerShell\Modules\HPEOneView.610\6.10.2778.2020\HPEOneView.610.psm1: 
                        line 8016
                        at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {}

VERBOSE: [SEND-OVREQUEST] Exception Message: The remote server returned an error: (412) Precondition Failed.
VERBOSE: [SEND-OVREQUEST] InnerException FullyQualifiedErrorId: 
VERBOSE: [SEND-OVREQUEST] InnerException Message: 
VERBOSE: [SEND-OVREQUEST] Exception
VERBOSE: [SEND-OVREQUEST] Getting Error Response
VERBOSE: [SEND-OVREQUEST] ERROR RESPONSE: {"errorSource":null,"data":{},"details":"The request contained an inva
lid X-API-Version header of 2800.  Only values 1 through 2400 are accepted.","message":"The specified API versio
n is not valid.","messageParameters":[],"nestedErrors":[],"errorCode":"INVALID_API_VERSION","recommendedActions"
:["Check the value of the X-API-Version header, then resend the request."]}
VERBOSE: [SEND-OVREQUEST] Response Status: HTTP 412 [Precondition Failed]
VERBOSE: [SEND-OVREQUEST] Response Header: Pragma = no-cache,no-cache
VERBOSE: [SEND-OVREQUEST] Response Header: Vary = Fri, 23 Jul 2021 01:15:10 GMT
VERBOSE: [SEND-OVREQUEST] Response Header: Content-Language = 0
VERBOSE: [SEND-OVREQUEST] Response Header: X-Frame-Options = Apache
VERBOSE: [SEND-OVREQUEST] Response Header: Strict-Transport-Security = timeout=15, max=100
VERBOSE: [SEND-OVREQUEST] Response Header: X-XSS-Protection = Keep-Alive
VERBOSE: [SEND-OVREQUEST] Response Header: X-Content-Type-Options = chunked
VERBOSE: [SEND-OVREQUEST] Response Header: Cache-Control = application/json; charset=UTF-8
VERBOSE: [SEND-OVREQUEST] Response Header: Date = 
VERBOSE: [SEND-OVREQUEST] Response Header: Expires = 
VERBOSE: [SEND-OVREQUEST] Response Header: Server = 
VERBOSE: [SEND-OVREQUEST] Response Header: Keep-Alive = 
VERBOSE: [SEND-OVREQUEST] Response Header: Connection = 
VERBOSE: [SEND-OVREQUEST] Response Header: Transfer-Encoding = 
VERBOSE: [SEND-OVREQUEST] Response Header: Content-Type = 
VERBOSE: [NEW-ERRORRECORD] Building ErrorRecord object
VERBOSE: [SEND-OVREQUEST] Cleaning up HttpWebRequest
VERBOSE: [CONNECT-OVMGMT] Exception caught when checking X-API version.
Connect-OVMgmt : [Send-OVRequest]: The specified API version is not valid. Check the value of the 
X-API-Version header, then resend the request.
At line:1 char:1
+ Connect-OVMgmt -Hostname XXXXXXXXX -UserName administrator -Passwor ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (Send-OVRequest:String) [Connect-OVMgmt], InvalidOperationExce 
   ption
    + FullyQualifiedErrorId : INVALID_API_VERSION,Connect-OVMgmt
 

PS C:\dev> $PSVersionTable | clip

PS C:\dev> $PSVersionTable 

Name                           Value                                                                           
----                           -----                                                                           
PSVersion                      5.1.19041.1023                                                                  
PSEdition                      Desktop                                                                         
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                         
BuildVersion                   10.0.19041.1023                                                                 
CLRVersion                     4.0.30319.42000                                                                 
WSManStackVersion              3.0                                                                             
PSRemotingProtocolVersion      2.3                                                                             
SerializationVersion           1.1.0.1                                                                         

PS C:\dev> 
</PRE>

**Comment by ChrisLynchHPE** [comment]
You must use the minimum version You are trying to use a newer build of the library with an older OneView appliance version.  This is not allowed.  The library has never supported taking a newer Major.Minor version of the library and using it with an older appliance.  You must use the 6.10 or "older" version of the library with your appliance.

**Reply by Cynomus** [reply]
Really wish the compatibility of the versions was more reliable, because most environments are going to be a mixture of system versions

**Reply by VincentBerger** [reply]
Please explain what makes you think this is not reliable. Like Chris explains, in an environment with multiple versions of OneView, you should use a version of the library that matches your oldest appliance, and it will work with newer appliances too.
